File: starter_kits/Java/hlt/Input.java
package hlt;

public class Input {
    private final String[] input;
    private int current;

    public Input(final String line) {
        input = line.split(" ");
    }

    public int getInt() {
        return Integer.parseInt(input[current++]);
    }

    public static Input readInput() {
        return new Input(readLine());
    }

    public static String readLine() {
        try {
            final StringBuilder builder = new StringBuilder();

            int buffer;
            for (; (buffer = System.in.read()) >= 0;) {
                if (buffer == '\n') {
                    break;
                }
                if (buffer == '\r') {
                    // Ignore carriage return if on windows for manual testing.
                    continue;
                }
                builder.append((char)buffer);
            }

            return builder.toString();
        } catch (final Exception e) {
            Log.log("Input connection from server closed. Exiting...");
            System.exit(0);
            throw new IllegalStateException(e);
        }
    }
}

File: starter_kits/Scala/hlt/Input.scala
package hlt

import scala.io.StdIn

object Input {
  def readInput = new Input(readLine)

  def readLine: String = try {
    StdIn.readLine()
  } catch {
    case e: Exception =>
      Log.log("Input connection from server closed. Exiting...")
      System.exit(0)
      throw new IllegalStateException(e)
  }
}

class Input(val line: String) {
  var input: Array[String] = line.split(" ")
  var current: Int = 0

  def getInt: Int = {
    val nextInt = input(current).toInt
    current = current + 1
    nextInt
  }
}
